swift/ql/lib/codeql/swift/dataflow/ExternalFlow.qll

```diff
@@ -263,14 +264,16 @@ module CsvValidation {
     )
   }
 
-  private string getInvalidModelKind() {
-    exists(string row, string kind | summaryModel(row) |
-      kind = row.splitAt(";", 8) and
-      not kind = ["taint", "value"] and
-      result = "Invalid kind \"" + kind + "\" in summary model."
-    )
+  private module KindValConfig implements SharedModelVal::KindValidationConfigSig {
+    predicate summaryKind(string kind) { summaryModel(_, _, _, _, _, _, _, _, kind, _) }
+
+    predicate sinkKind(string kind) { sinkModel(_, _, _, _, _, _, _, kind, _) }
+
+    predicate sourceKind(string kind) { sourceModel(_, _, _, _, _, _, _, kind, _) }
```

Comment:
I'm probably missing something here - isn't this circular?  We define the valid kinds to be all the kinds that are used in models, then we check the models all have valid kinds?

Reply:
I don't think this is circular. 🤔 The valid kinds are defined as any kinds listed in the `Valid[Sink|Source|Summary|Neutral]Kind` classes in [`ModelValidation.qll`](https://github.com/github/codeql/pull/13324/files#diff-91aa4933a4678a234e67334da0e0feb29f52cc2b3f396a2b13b5d63d64442db4R23), then we check in [`getInvalidModelKind`](https://github.com/github/codeql/pull/13324/files#diff-91aa4933a4678a234e67334da0e0feb29f52cc2b3f396a2b13b5d63d64442db4R161) if there exists a kind in any model that is not a valid kind.

Let me know if this makes sense or if you still think the logic is incorrect.

Reply:
... yeah, I thought this was supposed to compute the valid kinds, but it isn't, it's supposed to compute the used kinds which are then compared to the valid kinds where you linked to.